core/sail/federation/src/main/java/org/eclipse/rdf4j/sail/federation/evaluation/ParallelJoinCursor.java

```diff
@@ -75,11 +76,17 @@ public ParallelJoinCursor(EvaluationStrategy strategy,
 	public void run() {
 		evaluationThread = Thread.currentThread();
 		try {
-			while (!isClosed() && leftIter.hasNext()) {
-				CloseableIteration<BindingSet, QueryEvaluationException> evaluate = strategy.evaluate(rightArg,
-						leftIter.next());
-				toCloseList.add(evaluate);
-				rightQueue.put(evaluate);
+			while (true) {
+				synchronized (this) {
+					if (!closed && !isClosed() && leftIter.hasNext()) {
+						CloseableIteration<BindingSet, QueryEvaluationException> evaluate = strategy.evaluate(rightArg,
+								leftIter.next());
+						toCloseList.add(evaluate);
+						rightQueue.put(evaluate);
+					} else {
+						break;
+					}
+				}
```

Comment:
The problem was that between checking for closed and hasNext the iteration could have been closed. So I introduced some light locking in order to eliminate this race condition.